File: codes/frontend/src/lib/orgKeypair.test.ts
import { describe, it, expect } from 'vitest'
import {
  generateOrgKeypair,
  saveOrgKeypair,
  loadOrgKeypair,
  getOrCreateOrgKeypair,
  clearOrgKeypair,
  publicKeyFor,
  type KeyStore,
} from './orgKeypair'
import { signTicket, verifyTicket, createTicketPayload } from './ticket'

/** In-memory KeyStore so the libs are testable without a DOM. */
function memStore(): KeyStore & { dump: Map<string, string> } {
  const dump = new Map<string, string>()
  return {
    dump,
    getItem: (k) => (dump.has(k) ? dump.get(k)! : null),
    setItem: (k, v) => void dump.set(k, v),
    removeItem: (k) => void dump.delete(k),
  }
}

const POLL_A = '0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'
const POLL_B = '0xbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb'

describe('orgKeypair', () => {
  it('generates a keypair whose public key matches the private key', () => {
    const kp = generateOrgKeypair()
    expect(kp.privKey).toMatch(/^[0-9a-f]{64}$/)
    expect(kp.pubKey).toMatch(/^[0-9a-f]{64}$/)
    expect(publicKeyFor(kp.privKey)).toBe(kp.pubKey)
  })

  it('a generated key can sign a ticket that verifies against its public key', () => {
    const kp = generateOrgKeypair()
    const enc = signTicket(createTicketPayload(POLL_A, 1000), kp.privKey)
    expect(verifyTicket(enc, kp.pubKey, 1000).valid).toBe(true)
  })

  it('save then load returns the identical keypair', () => {
    const store = memStore()
    const kp = generateOrgKeypair()
    saveOrgKeypair(POLL_A, kp, store)
    expect(loadOrgKeypair(POLL_A, store)).toEqual(kp)
  })

  it('returns null for an unknown poll', () => {
    expect(loadOrgKeypair(POLL_A, memStore())).toBeNull()
  })

  it('getOrCreate is idempotent per poll and isolated across polls', () => {
    const store = memStore()
    const a1 = getOrCreateOrgKeypair(POLL_A, store)
    const a2 = getOrCreateOrgKeypair(POLL_A, store)
    const b1 = getOrCreateOrgKeypair(POLL_B, store)
    expect(a2).toEqual(a1) // same poll → same key
    expect(b1).not.toEqual(a1) // different poll → different key
  })

  it('returns null for a corrupt stored value without throwing', () => {
    const store = memStore()
    store.setItem('org-keypair-' + POLL_A.toLowerCase(), 'not json{')
    expect(loadOrgKeypair(POLL_A, store)).toBeNull()
    store.setItem('org-keypair-' + POLL_A.toLowerCase(), JSON.stringify({ privKey: 'short' }))
    expect(loadOrgKeypair(POLL_A, store)).toBeNull()
  })

  it('clear removes the stored key', () => {
    const store = memStore()
    const kp = generateOrgKeypair()
    saveOrgKeypair(POLL_A, kp, store)
    clearOrgKeypair(POLL_A, store)
    expect(loadOrgKeypair(POLL_A, store)).toBeNull()
  })
})

File: codes/mobile/test/core/crypto/org_keypair_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:zkvote_mobile/core/crypto/org_keypair.dart';
import 'package:zkvote_mobile/core/crypto/ticket.dart';

/// In-memory [KeyStore] so the lib is testable without platform storage.
class MemStore implements KeyStore {
  final Map<String, String> dump = {};
  @override
  String? getItem(String key) => dump[key];
  @override
  void setItem(String key, String value) => dump[key] = value;
  @override
  void removeItem(String key) => dump.remove(key);
}

const pollA = '0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa';
const pollB = '0xbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb';

void main() {
  final fixture = jsonDecode(
    File('test/fixtures/cross_client_vectors.json').readAsStringSync(),
  ) as Map<String, dynamic>;
  final ed = fixture['ed25519'] as Map<String, dynamic>;

  group('orgKeypair', () {
    test('publicKeyFor reproduces the golden TS public key', () {
      expect(publicKeyFor(ed['seedHex'] as String), ed['pubKey']);
    });

    test('generated keypair: hex-shaped and pubKey derives from privKey', () {
      final kp = generateOrgKeypair();
      expect(kp.privKey, matches(RegExp(r'^[0-9a-f]{64}$')));
      expect(kp.pubKey, matches(RegExp(r'^[0-9a-f]{64}$')));
      expect(publicKeyFor(kp.privKey), kp.pubKey);
    });

    test('a generated key signs a ticket that verifies against its pubKey', () {
      final kp = generateOrgKeypair();
      final enc = signTicket(createTicketPayload(pollA, 1000), kp.privKey);
      expect(verifyTicket(enc, kp.pubKey, now: 1000).valid, isTrue);
    });

    test('save then load returns the identical keypair', () {
      final store = MemStore();
      final kp = generateOrgKeypair();
      saveOrgKeypair(pollA, kp, store);
      expect(loadOrgKeypair(pollA, store), kp);
    });

    test('returns null for an unknown poll', () {
      expect(loadOrgKeypair(pollA, MemStore()), isNull);
    });

    test('getOrCreate is idempotent per poll and isolated across polls', () {
      final store = MemStore();
      final a1 = getOrCreateOrgKeypair(pollA, store);
      final a2 = getOrCreateOrgKeypair(pollA, store);
      final b1 = getOrCreateOrgKeypair(pollB, store);
      expect(a2, a1); // same poll → same key
      expect(b1 == a1, isFalse); // different poll → different key
    });

    test('corrupt stored value returns null without throwing', () {
      final store = MemStore();
      store.setItem('org-keypair-${pollA.toLowerCase()}', 'not json{');
      expect(loadOrgKeypair(pollA, store), isNull);
      store.setItem(
        'org-keypair-${pollA.toLowerCase()}',
        jsonEncode({'privKey': 'short'}),
      );
      expect(loadOrgKeypair(pollA, store), isNull);
    });

    test('clear removes the stored key', () {
      final store = MemStore();
      saveOrgKeypair(pollA, generateOrgKeypair(), store);
      clearOrgKeypair(pollA, store);
      expect(loadOrgKeypair(pollA, store), isNull);
    });
  });
}
